### PR TITLE
[v7r0] Add createClient decorator to add wrapper for all export_ functions automatically

### PR DESCRIPTION
**Pull request by andresailer**
This adds a ~~"metaclass"~~ decorator to automatically expose all the `export_` service functions from the client, which makes documentation for clients more complete (also ipython tab completion). Functions in the client that are explicitly defined are not overwritten. One limitation is that the argument names cannot be taken directly from the service function ~~(well maybe)~~, so only `*args` or nothing is defined (and **kwargs for url, timeout, rpc).

Example:
https://diracsailer.readthedocs.io/en/fixdoc/CodeDocumentation/WorkloadManagementSystem/Client/JobMonitoringClient.html

The changes to the client class source code are very limited

### CODE REVIEW

**Comment by fstagni** [comment]
Looks good to me, especially if it solves the "ipython completion". Of course it depends on the fact that in every *Client module we add the 3 magic lines.

**Comment by andresailer** [comment]
* I did some small tweaks to how the docstring for the functions look like, which show the type (though not very nicely.. comments welcome.) https://diracsailer.readthedocs.io/en/fixdoc/CodeDocumentation/WorkloadManagementSystem/Client/JobMonitoringClient.html#DIRAC.WorkloadManagementSystem.Client.JobMonitoringClient.JobMonitoringClient.getCounters
* use `six` to make the use of the metaclass python 2 and 3 compatible
* added the metaclass to all Clients inheriting from `Core.Base.Client` (except the old FTSClient)
* there is some side effect for the ReqClient (failing test). I have not yet understood how that goes wrong...

**Comment by andresailer** [comment]
Putting into WIP until there is time to discuss the broken test with @chaen

**Comment by andresailer** [comment]
Following the discussion with @chaen this morning: using a decorator instead of metaclass, and re-implemented without calling import lib but rather the `ast` module to get the functions from the service class. I dropped the adding of the argument type to the parameter, it looked terrible anyway...

(only the test for the documentation should fail now until #3860 is merged)

**Comment by fstagni** [comment]
I have 2 feature requests:
- support extensions (of services)
- there's a standard way to refer to services, and it's in the form of "System/Service". So, it would be nice to see `@createClient("DataManagement/DataIntegrity")

**Comment by andresailer** [comment]
Will implement the requested features in the following way:
Use the "SystemName/ServiceName" to infer the location of the service implementation file in "$DIRAC/DIRAC/SystemNameSystem/Service/ServiceNameHandler.py", and in addition look into all the folders in
`$DIRAC/*DIRAC/` (e.g. ILCDIRAC, LHCbDIRAC...) to see if there are extensions. Extensions are parsed first, so if functions are added in extensions they won't get overwritten. Obtaining the list of extensions via the configuration would add too many side effects to this.

The overhead from all this is small as it is only done the first time a client module is imported, though one could add an environment variable to short cut all of this.

I'll also add some more documentation to the createClient implementation

**Comment by andresailer** [comment]
Implemented the feature requests of @fstagni

File: TransformationSystem/Client/TransformationClient.py [hunk 1]
```diff
@@ -3,13 +3,14 @@
...
+from DIRAC.Core.Base.Client import Client, createClient
 from DIRAC.Core.Utilities.List                                     import breakListIntoChunks
 from DIRAC.ConfigurationSystem.Client.Helpers.Operations           import Operations
 
-class TransformationClient( Client ):
 
-  """ Exposes the functionality available in the DIRAC/TransformationHandler
+@createClient('TransformationSystem/TransformationManager')
+class TransformationClient(Client):
+  """ Exposes the functionality available in the DIRAC/TransformationManagerHandler
```
Comment (fstagni) [inline TransformationSystem/Client/TransformationClient.py]: You can review (remove now?) the content of the docstring below this line

**Comment by andresailer** [comment]
I thought I had added it to all Clients that inherit from Core.Base.Client, but I missed a few.
I did not want to touch FTSClient as it will disappear soon, right?
There are more clients (*/Client/*Client.py) that do not inherit from Core.Base.Client.
E.g.:
```
DIRAC (v7_metaClients)$ find . -wholename "*/Client/*Client.py" -exec grep -nH "class .*Client *(" {} \; | grep object
./MonitoringSystem/Client/MonitoringClient.py:11:class MonitoringClient(object):
./FrameworkSystem/Client/ProxyManagerClient.py:22:class ProxyManagerClient( object ):
./FrameworkSystem/Client/MonitoringClient.py:56:class MonitoringClient( object ):
./FrameworkSystem/Client/SecurityLogClient.py:12:class SecurityLogClient(object):
./AccountingSystem/Client/DataStoreClient.py:21:class DataStoreClient(object):
./WorkloadManagementSystem/Client/SandboxStoreClient.py:25:class SandboxStoreClient(object):
./WorkloadManagementSystem/Client/WMSClient.py:22:class WMSClient(object):
./ResourceStatusSystem/Client/ResourceStatusClient.py:18:class ResourceStatusClient(object):
./ResourceStatusSystem/Client/ResourceManagementClient.py:18:class ResourceManagementClient(object):
```

**Comment by fstagni** [comment]
Good. Others should not be touched. I let @chaen the final review, for me this is OK.

File: TransformationSystem/Client/TransformationClient.py (hunk 1 above)
Comment (andresailer) [reply TransformationSystem/Client/TransformationClient.py]: Done

**Comment by andresailer** [comment]
Actually the transformation Client doesn't work like the others, because the "Base" service class contains all the functions
https://github.com/DIRACGrid/DIRAC/blob/29591a5b1818cc48854e845e151e92ba84a4af88/TransformationSystem/Service/TransformationManagerHandler.py#L19
and the usually named class doesn't.
https://github.com/DIRACGrid/DIRAC/blob/29591a5b1818cc48854e845e151e92ba84a4af88/TransformationSystem/Service/TransformationManagerHandler.py#L713

And we don't pick up functions from base classes, yet.

**Comment by fstagni** [comment]
I don't remember (and I don't see) the reason why the TransformationManager has that type of inheritance...?

**Comment by andresailer** [comment]
So I merge the two and rename `TransformationManagerHandlerBase` to `TransformationManagerHandler`?

**Comment by fstagni** [comment]
I think it would make sense, thanks.

**Pull request merged**